### ISSUE TITLE
slow performance of banded LU factorisation

### ISSUE THREAD

**Issue by gasagna** [xianyi/OpenBLAS#901]
This issue has been observed under Julia.  Please see JuliaLang/julia#16653 for details.

**Comment by brada4** [comment]
Works like a charm in generic setup. Unlikely OpenBLAS which takes 5% of run time is at fault at (non-lapack) gbtrf regression
```
julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.3 (2016-01-12 21:37 UTC)
 _/ |\__'_|_|_|\__'_|  |  
|__/                   |  x86_64-suse-linux

julia> versioninfo()
Julia Version 0.4.3
Commit a2f713d* (2016-01-12 21:37 UTC)
Platform Info:
  System: Linux (x86_64-suse-linux)
  CPU: AMD A10-5750M APU with Radeon(tm) HD Graphics
  WORD_SIZE: 64
  BLAS: libopenblas (DYNAMIC_ARCH NO_AFFINITY Piledriver)
  LAPACK: liblapack
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
julia> import Base.LAPACK.gbtrf!
julia> for n in 10:20
        data = randn(3n+1, 10000)
        println(n, " ", minimum([@elapsed gbtrf!(n, n, 10000, data) for i = 1:10]))
       end
10 0.021106314
11 0.020081044
12 0.019893869
13 0.022235508
14 0.022406199
15 0.022024028
16 0.022753171
17 0.022580523
18 0.022759876
19 0.022535825
20 0.023108524
julia> 
```

**Comment by gasagna** [comment]
What machine are you on? your times are ten time slower then mines. However, they are comparable with what I get on another linux machine.

````julia
julia> versioninfo()
Julia Version 0.4.5
Commit 2ac304d (2016-03-18 00:58 UTC)
Platform Info:
  System: Linux (x86_64-unknown-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-2630 v3 @ 2.40GHz
  WORD_SIZE: 64
  BLAS: libopenblas (DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas
  LIBM: libm
  LLVM: libLLVM-3.3

julia> for n in 10:20
          data = randn(3n+1, 10000)
          println(n, " ", minimum([@elapsed gbtrf!(n, n, 10000, data) for i = 1:10]))
       end
10 0.020331116
11 0.020589112
12 0.022374969
13 0.026552423
14 0.026454015
15 0.026838984
16 0.026198325
17 0.026692526
18 0.028327944
19 0.028919155
20 0.029936612
````

On my macbook i get, (using Apple's blas)

````julia
julia> versioninfo()
Julia Version 0.5.0-dev+4422
Commit d07b849* (2016-05-31 20:46 UTC)
Platform Info:
  System: Darwin (x86_64-apple-darwin15.5.0)
  CPU: Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz
  WORD_SIZE: 64
  BLAS: libgfortblas
  LAPACK: liblapack
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1 (ORCJIT, haswell)

julia> for n in 10:20
           data = randn(3n+1, 10000)
           println(n, " ", minimum([@elapsed gbtrf!(n, n, 10000, data) for i = 1:10]))
       end
10 0.000838607
11 0.000924276
12 0.00089478
13 0.001032492
14 0.001140993
15 0.001233413
16 0.00111085
17 0.001400441
18 0.001475315
19 0.001660101
20 0.001415923
````
which is **much** faster! With the default openblas setup shipped with julia I get:

````julia
julia> versioninfo()
Julia Version 0.4.5
Commit 2ac304d* (2016-03-18 00:58 UTC)
Platform Info:
  System: Darwin (x86_64-apple-darwin15.4.0)
  CPU: Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz
  WORD_SIZE: 64
  BLAS: libopenblas (DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3

julia> for n in 10:20
           data = randn(3n+1, 10000)
           println(n, " ", minimum([@elapsed gbtrf!(n, n, 10000, data) for i = 1:10]))
       end
10 0.001943523
11 0.001969065
12 0.002187122
13 0.002368863
14 0.002571177
15 0.002755816
16 0.210919274
17 0.203663198
18 0.242170501
19 0.230052181
20 0.241873166
````
where time jumps by two orders of magnitude.

**Comment by brada4** [comment]
It looks like thermal throttling.

**Comment by gasagna** [comment]
Why would it behave like this?

![figure_1](https://cloud.githubusercontent.com/assets/744738/15740601/6bbe90f0-28ad-11e6-9ebf-3858c6394eb8.png)

**Comment by andreasnoack** [comment]
I also see this problem on my Mac but only for the Nehalem and Haswell kernels. It's not and issue with Sandybridge. @gasagna Try lauching Julia with

```jl
OPENBLAS_CORETYPE=Sandybridge ./julia
```

I tried on a Haswell running Linux and I didn't see the problem. The problem doesn't seem to be related to threading.

**Comment by martin-frbg** [comment]
Apple- and Haswell specific issue - could it be an alignment thing like the recent JuliaLang/julia#5728 ?
(I notice at least dscal_microk_haswell still has the ".align 16" that were dropped from gemv in commit 6e7be06 with the reasoning that it should read ".align 4" to achieve the same on OSX and was not needed anyway)

**Comment by gasagna** [comment]
Is there any update on this issue?

**Comment by martin-frbg** [comment]
Core developers appear to be busy elsewhere. In the absence of better suggestions I guess you could
try my hunch of removing the two ".align 16" lines from kernel/x86_64/dscal_microk_haswell-2.c and see if this brings any improvement. (I do not have a Darwin system to test this)

**Comment by gasagna** [comment]
Thanks. Will have a try at this asap and report here.

**Comment by martin-frbg** [comment]
@gasagna one week on, did you get a chance to test this ? (If only to keep me from offering similarly silly ideas in vaguely related cases...)

**Comment by gasagna** [comment]
Nope, sorry. Been busy with work. Feel free to investigate further...

**Comment by martin-frbg** [comment]
So, four months later, anybody with Julia on a Mac and time to do the timings without the two .align 16 directives in /dscal_microk_haswell-2.c ?

**Comment by martin-frbg** [comment]
@gasagna could you check your build options for OpenBLAS if you have the time ? As noticed (and fixed) only very recently, the default "-O2" from COMMON_OPT was not passed to the fortran compiler, so LAPACK functions built as part of OpenBLAS would be completely unoptimized unless one happened to preset the FFLAGS environment variable or edit Makefile.system directly before compiling. From the julia ticket it seems you were using OpenBLAS with LAPACK built in (LAPACK: libopenblas64_ in the initial log you posted) so it seems a distinct possibility that this might trivially explain the slowness.

**Comment by martin-frbg** [comment]
Presumed fixed by JuliaLang/julia#1471
